### PR TITLE
Add the support of "MultipleHubs" in KlusterletConfig.

### PR DESCRIPTION
**Pull request by xuezhaojun**
Fixes: https://issues.redhat.com/browse/ACM-11114

### CODE REVIEW

**Comment by xuezhaojun** [comment]
cc @clyang82

File: pkg/bootstrap/render.go [hunk 1]
```diff
@@ -319,6 +337,39 @@ func (b *KlusterletManifestsConfig) Generate(ctx context.Context, clientHolder *
...
+			multipleHubsEnabled = true
+			for _, s := range b.klusterletconfig.Spec.BootstrapKubeConfigs.LocalSecrets.KubeConfigSecrets {
+				ns := os.Getenv(constants.PodNamespaceEnvVarName)
+				secret, err := clientHolder.KubeClient.CoreV1().Secrets(ns).Get(ctx, s.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, err
+				}
+				bootstrapKubeConfigSecrets = append(bootstrapKubeConfigSecrets, BootstrapKubeConfigSecret{
+					Name:       s.Name,
+					KubeConfig: base64.StdEncoding.EncodeToString(secret.Data["kubeconfig"]),
```
Comment (zhujian7) [inline pkg/bootstrap/render.go:364]: should we check whether the "kubeconfig" key exists in the secret data?

File: pkg/bootstrap/render.go [hunk 2]
```diff
@@ -319,6 +337,39 @@ func (b *KlusterletManifestsConfig) Generate(ctx context.Context, clientHolder *
...
+			registrationConfiguration.FeatureGates = append(registrationConfiguration.FeatureGates,
+				operatorv1.FeatureGate{
+					Feature: string(apifeature.MultipleHubs),
+					Mode:    operatorv1.FeatureGateModeTypeEnable,
+				})
+			registrationConfiguration.BootstrapKubeConfigs = b.klusterletconfig.Spec.BootstrapKubeConfigs
+
+			multipleHubsEnabled = true
+			for _, s := range b.klusterletconfig.Spec.BootstrapKubeConfigs.LocalSecrets.KubeConfigSecrets {
+				ns := os.Getenv(constants.PodNamespaceEnvVarName)
```
Comment (zhujian7) [inline pkg/bootstrap/render.go:357]: The secret can only be in the pod namespace? in MCE it is `multicluster-engine`?

File: pkg/controller/importconfig/manager.go [hunk 3]
```diff
@@ -123,6 +123,19 @@ func Add(ctx context.Context,
 					UpdateFunc:  func(e event.UpdateEvent) bool { return true },
 				})),
 		).
+		Watches(
+			&corev1.Secret{},
```
Comment (zhujian7) [inline pkg/controller/importconfig/manager.go:127]: will this watch all secrets in the hub cluster?

File: pkg/controller/importconfig/klusterletconfighandler.go [hunk 4]
```diff
@@ -69,3 +71,81 @@ func IndexManagedClusterByKlusterletconfigAnnotation(obj interface{}) ([]string,
...
+	evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	e.enqueue(evt.Object.GetName(), q)
+}
+
+func (e *enqueueManagedClusterByBootstrapKubeConfigSecrets) Generic(ctx context.Context,
+	evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+	e.enqueue(evt.Object.GetName(), q)
+}
+
+func (e *enqueueManagedClusterByBootstrapKubeConfigSecrets) enqueue(secretName string, q workqueue.RateLimitingInterface) {
```
Comment (zhujian7) [inline pkg/controller/importconfig/klusterletconfighandler.go:111]: can we have some rules like the secret must be in the managed cluster namespace, so here we can enqueue the secret.namespace directly?

File: pkg/bootstrap/render.go (hunk 2 above)
Comment (xuezhaojun) [reply pkg/bootstrap/render.go:357]: Yes, within the same namespace of the pod.

File: pkg/bootstrap/render.go (hunk 1 above)
Comment (xuezhaojun) [reply pkg/bootstrap/render.go:364]: Sure, the check added.

File: pkg/controller/importconfig/manager.go (hunk 3 above)
Comment (xuezhaojun) [reply pkg/controller/importconfig/manager.go:127]: Good point, change to watch only secrets under pod namespace.

File: pkg/controller/importconfig/klusterletconfighandler.go (hunk 4 above)
Comment (xuezhaojun) [reply pkg/controller/importconfig/klusterletconfighandler.go:111]: Not quit get the idea of enqueue the secret.namespace, may need to discuss offline.

File: pkg/bootstrap/render.go [hunk 5]
```diff
@@ -319,6 +337,45 @@ func (b *KlusterletManifestsConfig) Generate(ctx context.Context, clientHolder *
...
+				secret, err := clientHolder.KubeClient.CoreV1().Secrets(ns).Get(ctx, s.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, err
```
Comment (qiujian16) [inline pkg/bootstrap/render.go:360]: do we really want to return error here? what it type is not local secrets and the secrets are specified here?

**Comment by xuezhaojun** [comment]
whether “watch all secrets in mce ns” will cause a performance issue should be considered.

File: pkg/bootstrap/render.go [hunk 6]
```diff
@@ -319,6 +337,45 @@ func (b *KlusterletManifestsConfig) Generate(ctx context.Context, clientHolder *
 		renderConfig.ImagePullSecretConfig = imagePullSecretConfig
 	}
 
+	// MultipleHubs
+	// Doesn't affect on the local-cluster.
+	// Using MultipleHubs can controls the bootstrap kubeconfig secret/secrets easier.
+	multipleHubsEnabled := false
+	bootstrapKubeConfigSecrets := []BootstrapKubeConfigSecret{}
+	if local, ok := b.ManagedClusterLabels[constants.SelfManagedLabel]; !ok || local != "true" {
```
[the remark this answers is not in the document]
Comment (xuezhaojun) [reply pkg/bootstrap/render.go]: Update.

File: pkg/bootstrap/render.go [hunk 7]
```diff
@@ -319,6 +337,45 @@ func (b *KlusterletManifestsConfig) Generate(ctx context.Context, clientHolder *
 		renderConfig.ImagePullSecretConfig = imagePullSecretConfig
 	}
 
+	// MultipleHubs
+	// Doesn't affect on the local-cluster.
+	// Using MultipleHubs can controls the bootstrap kubeconfig secret/secrets easier.
+	multipleHubsEnabled := false
```
[the remark this answers is not in the document]
Comment (xuezhaojun) [reply pkg/bootstrap/render.go:344]: I don't quit understand the idea, the multiplehubs need to render the `Klusterlet`, it's not likely work like `GenerateKlusterletCRDsV1` which only need to render independent files.

File: pkg/bootstrap/render.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xuezhaojun) [reply pkg/bootstrap/render.go]: Updated.

File: pkg/bootstrap/render.go (hunk 5 above)
Comment (xuezhaojun) [reply pkg/bootstrap/render.go:352]: So if type is not localsecrets the path would go down here.

File: pkg/controller/importconfig/manager.go [hunk 8]
```diff
@@ -123,6 +128,27 @@ func Add(ctx context.Context,
 					UpdateFunc:  func(e event.UpdateEvent) bool { return true },
 				})),
 		).
+		Watches(
```
Comment (zhujian7) [inline pkg/controller/importconfig/manager.go:131]: Please also check if the import-controller serviceaccount in the backplane-operator has the permission.

**Comment by qiujian16** [comment]
need rebase, otherwise looks good.

File: pkg/controller/importconfig/manager.go (hunk 8 above)
Comment (xuezhaojun) [reply pkg/controller/importconfig/manager.go:131]: Good point, checked: https://github.com/stolostron/backplane-operator/blob/0d2edc3cf537a2eb4c0c9af71e1a98d104110ca0/pkg/templates/charts/toggle/server-foundation/templates/managedcluster-import-role.yaml#L73C1-L86C10

**Comment by xuezhaojun** [comment]
> need rebase, otherwise looks good.

Thanks! Code rebased.

**Comment by xuezhaojun** [comment]
@qiujian16 @zhujian7  All checks are passed, looking for another round of reivew, thanks!

**Comment by qiujian16** [comment]
/approve
@zhujian7 would you do the final review?

**Comment by zhujian7** [comment]
/lgtm
/approve
@xuezhaojun please unhold if QE can contain this in this train.

**Pull request merged**